arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/raise/RaiseAccumulate.kt

```diff
@@ -772,4 +766,47 @@ public open class RaiseAccumulate<Error>(
     }
     return block(raise)
   }
+
+  @PublishedApi internal fun addErrors(newErrors: Iterable<Error>) { errors.addAll(newErrors) }
+  @PublishedApi internal fun hasErrors(): Boolean = errors.isNotEmpty()
+  @PublishedApi internal fun raiseErrors(): Nothing = raise.raise(errors.toNonEmptyListOrNull()!!)
+
+  public fun <A> Either<Error, A>.bindOrAccumulate(): Value<A> =
+    accumulating { this@bindOrAccumulate.bind() }
+
+  public fun <A> Iterable<Either<Error, A>>.bindAllOrAccumulate(): Value<List<A>> =
+    accumulating { this@bindAllOrAccumulate.bindAll() }
+
+  public fun <A> EitherNel<Error, A>.bindNelOrAccumulate(): Value<A> =
+    accumulating { this@bindNelOrAccumulate.bindNel() }
+
+  public fun ensureOrAccumulate(condition: Boolean, raise: () -> Error) {
+    accumulating { ensure(condition, raise) }
+  }
+
+  public fun <B: Any> ensureNotNullOrAccumulate(value: B?, raise: () -> Error) {
+    ensureOrAccumulate(value != null, raise)
+  }
+
+  public inline fun <A> accumulating(block: RaiseAccumulate<Error>.() -> A): Value<A> =
+    recover(inner@{
+      Ok(block(RaiseAccumulate(this@inner)))
+    }) {
+      addErrors(it)
+      Error()
+    }
+
+  public abstract inner class Value<out A> {
+    public abstract val result: A
+    public operator fun getValue(value: Nothing?, property: KProperty<*>): A = result
```

Comment:
Do all these need to be `public`?

Reply:
`getValue` has to be public for `by` to work. I added `value` just to allow using the `Value` on its own (some people don't like delegation), but it can be removed.

Reply:
I would initially hide whatever is possible, we can always make it public afterwards.

Reply:
Done!